src/crypto/sha256sum.cpp

```diff
@@ -0,0 +1,90 @@
+#include "crypto/sha256sum.h"
+
+#include <algorithm>       // min
+#include <cstdio>          // BUFSIZ
+#include <fstream>         // ifstream
+#include <ios>             // ios_base::{binary, in, ate}
+#include <openssl/evp.h>
+
+#include "misc_log_ex.h"
+
+#undef MONERO_DEFAULT_LOG_CATEGORY
+#define MONERO_DEFAULT_LOG_CATEGORY "util"
+
+namespace {
+	constexpr size_t SHA256_MD_SIZE = 32;
+	static_assert(sizeof(crypto::hash::data) >= SHA256_MD_SIZE, "crypto::hash::data is too small for SHA-256");
+}
+
+namespace tools
+{
+	bool sha256sum(const uint8_t *data, size_t len, crypto::hash &hash) noexcept
+	{
+		if (!EVP_Digest(data, len, (unsigned char*) hash.data, NULL, EVP_sha256(), NULL)) {
+			MERROR("Failed to digest with OpenSSL SHA-256 EVP");
+			return false;
+		}
+
+		return true;
+	}
+
+	bool sha256sum(const std::string &filename, crypto::hash &hash) noexcept
+	{
+		// Open file for reading in binary mode, immediately putting the cursor at the end
+		std::ifstream f(filename, std::ios_base::binary | std::ios_base::in | std::ios::ate);
+		if (!f)
+		{
+			MERROR("sha256sum: Could not open file '" << filename << "' for reading");
+			return false;
+		}
+
+		// Retreive file size and go back to the beginning of the stream
+		size_t size_left = f.tellg();
+		f.seekg(0, std::ios::beg);
+
+		// Initialize digest context
+		EVP_MD_CTX* ctx;
+		if (NULL == (ctx = EVP_MD_CTX_new()))
+		{
+			MERROR("Failed to create new OpenSSL EVP_MD_CTX");
+			return false;
+		}
+		else if (!EVP_DigestInit_ex(ctx, EVP_sha256(), NULL))
+		{
+			MERROR("Failed to initialize OpenSSL SHA-256 EVP context");
+			return false;
```

Comment:
This and all subsequent error returns will leak the ctx.

Reply:
No, it won't leak. `ctx` is an `std::unique_ptr` instance which will call `evp_md_ctx_freer` in destructor.